src/storage/src/source/mod.rs

```diff
@@ -806,169 +866,640 @@ where
+                        unconsumed_partitions,
+                        source_upper.clone(),
+                    ));
+
+                    last_tick_time = now;
                 }
             }
         }));
 
         let activator = scope.activator_for(&info.address[..]);
         move |cap_set, output| {
-            if !active {
-                // Empty out the `CapabilitySet` to indicate that we're done.
-                cap_set.downgrade(&[]);
-                return;
-            }
-
-            // Accumulate updates to bytes_read for Prometheus metrics collection
-            let mut bytes_read = 0;
-            // Accumulate updates to offsets for system table metrics collection
-            let mut metric_updates = HashMap::new();
             // Record operator has been scheduled
+            // WIP: Should we have these metrics for all three involved
+            // operators?
             source_metrics.operator_scheduled_counter.inc();
 
             let mut context = Context::from_waker(&waker);
 
             let timer = Instant::now();
 
-            while let Poll::Ready(Some(event)) = source_reader.as_mut().poll_next(&mut context) {
-                match event {
-                    Event::Progress(upper) => {
-                        let ts = upper.as_option().cloned().unwrap_or(Timestamp::MAX);
-                        for partition_metrics in source_metrics.partition_metrics.values_mut() {
-                            partition_metrics.closed_ts.set(ts);
-                        }
+            // We just use an advancing number for our capability. No one cares
+            // about what this actually is, downstream.
+            let mut batch_counter = 0;
+
+            while let Poll::Ready(Some(update)) = source_reader.as_mut().poll_next(&mut context) {
+                let (messages, non_definite_errors, unconsumed_partitions, source_upper) =
+                    if let Some(update) = update {
+                        update
+                    } else {
+                        trace!("source_reader({id}) {worker_id}/{worker_count}: is terminated");
+                        // We will never produce more data, clear our capabilities to
+                        // communicate this downstream.
+                        cap_set.downgrade(&[]);
+                        return;
+                    };
+
+                trace!(
+                    "create_source_raw({id}) {worker_id}/{worker_count}: message_batch.len(): {:?}",
+                    messages.len()
+                );
+                trace!(
+                    "create_source_raw({id}) {worker_id}/{worker_count}: source_upper: {:?}",
+                    source_upper
+                );
+
+                // We forward only the partitions that we are responsible for to
+                // the remap operator.
+                let source_upper_summary = SourceUpperSummary {
+                    source_upper: source_upper.clone(),
```

Comment:
where is this `source_upper` filtered to the specific partitions?

Reply:
It's not filtered because the source only emits an upper containing the partitions that it is responsible for. We could assert that here, though.